sympy/matrices/matrices.py

```diff
@@ -38,6 +39,31 @@ def _iszero(x):
         return None
 
 
+def _is_probably_zero(x):
+    """Tests by plugging in random rational numbers. Returns True if several
+    substitution resulted in zero, otherwise returns False. Does not attempt
+    floating point evaluation or any symbolic simplification, thus returns
+    quickly. Suitable for polynomials and rational functions because, for
+    example, ``x*(x+1) - x - x**2`` evaluates to zero when any rational number
+    is plugged in."""
+    try:
+        is_zero = x.is_zero
+    except AttributeError:
+        is_zero = None
+    if is_zero is None:
+        variables = x.free_symbols
+        if isinstance(variables, set) and len(variables) > 0:
+            for _ in range(5):
+                substitutions = {v: random_complex_number(-5, 0, 5, 0, rational=True,
+                    tolerance=0.001) for v in variables}
+                if x.xreplace(substitutions) != 0:
+                    is_zero = False
+                    break
+            if is_zero is None:
+                is_zero = True
```

Comment:
How about using random integers? Also I think we can add `is_constant_zero` method to the `Expr`.

Reply:
I'd be worried about something like `sin(pi*x)` if x was only tested at integers. Testing at five random rationals with 3-digit denominators gives me a lot more confidence that the expression does not just evaluate to 0 by coincidence. 

Moving it up to Expr level is a reasonable thing to consider, but I have to repeat that this isn't really a zero-testing method for general expressions. It will not detect `sin(x)**2 + cos(x)**2 - 1` as being zero, for example. It only works when x, being replaced by a rational number, gets automatically combined with surrounding rational numbers, collapsing the formula into zero. So I'm afraid it would be false advertisement to add it to Expr as a general zero tester.

I think there is a reason why this module already had a function called `is_zero`, and why `find_reasonable_pivot` provided a "zerofunc" parameter, allowing for other such functions. It's that the linear algebra manipulations have this specific need for testing many potential pivots for being zero, without necessarily obtaining a proof of them being zero.

Reply:
> we can add is_constant_zero method to the Expr

We have `is_constant` and `equals` already. (If we find use elsewhere then maybe a method like `rational_zero(den_digits=3, repeat=5)` which raises an error if anything other than a rational expression in the variables is being tested. For now, I agree with @normalhuman 's sentiments and would vote to leave it only here.)

Reply:
> How about using random integers or rationals generated from random integers

@ylemkimon , are you suggesting (as it appears to me) that using `nsimplify` in `random_complex` is overkill for what is happening? I think a more meaningful approach would be to generate the random numbers then use Rational and, optionally, make a call to the `limit_denominator` method (changing `tolerance` to `maximum_denominator`): e.g. `Random(random()).limit_denominator(1000)`. On the other hand, rather than making rational deal with the large denominators in the first place, `Rational(randint(-5000, 5000), 1000)` ... and maybe that's what you meant by your comment?

Reply:
`Rational(randint(-5000, 5000), 1000)`  would be always zero for `sin(1000*pi*x)`. Having fixed denominator is making the test less trustworthy. The original version of this PR had `Rational(randint(...) / randint(...))` but I found this approach awkward because there is a good chance of canceling the numerator and denominator. 

`nsimplify` may be expensive in general, but it shouldn't be with `rational=True`, when it's essentially a call to Python's built-in `fractions.Fraction(p, q).limit_denominator(n)`

Reply:
Doh...another freshman flounder. Also, I think I looked too quickly at the nsimplify function.What you have done make sense to me and it’s pretty straightforward.  Any objections to committing?

Reply:
@smichr @normalhuman 
> We have `is_constant` and `equals` already.

> Moving it up to Expr level is a reasonable thing to consider, but I have to repeat that this isn't really a zero-testing method for general expressions.

Yes, but `equals` involves expensive `simplify` and `is_constant` just checks for whether it's a constant. I think as zero-check is used in many places, we can have special zero-check method, which implements various zero-checking methods including this one, with `method` argument or `quick` argument to control what method to use.

> are you suggesting (as it appears to me) that using nsimplify in random_complex is overkill for what is happening?

Yes, I thought just calling `randint` with fixed denominator, e.g., `10**n` or some large prime number, would suffice and be faster, but now I think, as @normalhuman suggested:

> Having fixed denominator is making the test less trustworthy.

Reply:
The last thing I want to think about is the possibility of random failures as was mentioned.